Sources/ApolloTestSupport/AsyncResultObserver.swift

```diff
@@ -0,0 +1,58 @@
+import XCTest
+
+public class AsyncResultObserver<Success, Failure> where Failure: Error {
+  public typealias ResultHandler = (Result<Success, Failure>) throws -> Void
+  
+  private class AsyncResultExpectation: XCTestExpectation {
+    let file: StaticString
+    let line: UInt
+    let handler: ResultHandler?
+    
+    init(description: String, file: StaticString = #filePath, line: UInt = #line, handler: ResultHandler?) {
+      self.file = file
+      self.line = line
+      self.handler = handler
+
+      super.init(description: description)
+    }
+  }
+  
+  private let testCase: XCTestCase
+  private var expectations: [AsyncResultExpectation] = []
+  
+  public init(testCase: XCTestCase) {
+    self.testCase = testCase
+  }
+  
+  public func expectation(description: String, file: StaticString = #filePath, line: UInt = #line, resultHandler: ResultHandler? = nil) -> XCTestExpectation {
+    let expectation = AsyncResultExpectation(description: description, file: file, line: line, handler: resultHandler)
+    expectation.assertForOverFulfill = true
+    
+    expectations.append(expectation)
+    
+    return expectation
+  }
+  
+  public func handler(_ result: Result<Success, Failure>) {
```

Comment:
Not really. The `handler` method on the observer is a normal completion handler that you can pass to any function that takes one (like `fetch` or `watch`). You set up expectations on the observer (which fills the array), and every time the handler is called it will take just the first expectation and fulfill it (after running the expectation's `handler`). It will only remove an expectation from the array after its fulfillment count is reached. That means you can set up consecutive expectations, which is useful is you expect a `fetch` to be called first with a result from the cache and then with a result from the server for example.

Reply:
ah ok so it's more of a convenience method to be able to pass as a lambda?

Reply:
Yep, the whole observer encapsulates common test patterns for async results, and the `handler` method is the entry point that you use to hook it up to a function that expects a completion handler. (I actually tried using `callAsFunction` because I was hoping that would allow you to pass the observer directly, but unfortunately that doesn't work.)